### PR TITLE
More new agent work

### PR DESCRIPTION
**Pull request by trask**

### CODE REVIEW

File: agent/src/main/java/com/microsoft/applicationinsights/agent/internal/MainEntryPoint.java [hunk 1]
```diff
@@ -1,20 +1,23 @@
...
+import com.microsoft.applicationinsights.agent.internal.config.AgentBuiltInConfiguration;
+import com.microsoft.applicationinsights.agent.internal.config.AgentConfiguration;
+import com.microsoft.applicationinsights.agent.internal.config.ClassInstrumentationData;
+import com.microsoft.applicationinsights.agent.internal.config.MethodInfo;
+import com.microsoft.applicationinsights.agent.internal.config.builder.XmlAgentConfigurationBuilder;
 import com.microsoft.applicationinsights.agent.internal.utils.Global;
-import org.glowroot.xyzzy.engine.config.InstrumentationDescriptor;
-import org.glowroot.xyzzy.engine.config.InstrumentationDescriptors;
+import com.microsoft.applicationinsights.internal.logger.InternalLogger;
+import org.glowroot.xyzzy.engine.config.*;
```
Comment (littleaj) [inline agent/src/main/java/com/microsoft/applicationinsights/agent/internal/MainEntryPoint.java]: style check flags this, explicit imports only

File: agent/src/main/java/com/microsoft/applicationinsights/agent/internal/MainEntryPoint.java [hunk 2]
```diff
@@ -1,20 +1,23 @@
...
 import org.slf4j.Logger;
 
 import java.io.File;
+import java.io.IOException;
 import java.lang.instrument.Instrumentation;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (littleaj) [inline agent/src/main/java/com/microsoft/applicationinsights/agent/internal/MainEntryPoint.java]: here too

File: agent/src/main/java/com/microsoft/applicationinsights/agent/internal/MainEntryPoint.java [hunk 3]
```diff
@@ -46,21 +49,144 @@ public static void premain(Instrumentation instrumentation, File agentJarFile) {
...
+                            if (!signature.equals(ClassInstrumentationData.ANY_SIGNATURE_MARKER)) {
+                                // TODO parse signature and call addAllMethodParameterTypes() and methodReturnType()
+                                InternalLogger.INSTANCE.warn("signature attribute is not currently supported");
+                            }
+
+                            adviceConfigs.add(builder
+                                    // xyzzy doesn't support threshold, so threshold is embedded into message and
+                                    // then parsed out by the agent to decide whether to report telemetry
+                                    .spanMessageTemplate(
+                                            "{{className}}.{{methodName}}#" + classInstrumentationData.getClassType() +
```
Comment (littleaj) [inline agent/src/main/java/com/microsoft/applicationinsights/agent/internal/MainEntryPoint.java]: does `spanMessageTemplate` validate its input?

File: agent/src/main/java/com/microsoft/applicationinsights/agent/internal/MainEntryPoint.java [hunk 4]
```diff
@@ -46,21 +49,144 @@ public static void premain(Instrumentation instrumentation, File agentJarFile) {
...
+
+        InstrumentationDescriptor instrumentationDescriptor = buildCustomInstrumentation(agentConfiguration);
+        if (instrumentationDescriptor != null) {
+            // need to test before enabling this
+            // instrumentationDescriptors.add(instrumentationDescriptor);
+        }
+        return instrumentationDescriptors;
+    }
+
+    private static InstrumentationDescriptor buildCustomInstrumentation(AgentConfiguration agentConfiguration) {
```
Comment (littleaj) [inline agent/src/main/java/com/microsoft/applicationinsights/agent/internal/MainEntryPoint.java]: Are there tests for this scenario? if not, let's create an issue/task

File: test/smoke/framework/testCore/src/main/java/com/microsoft/applicationinsights/smoketest/docker/AiDockerClient.java [hunk 5]
```diff
@@ -81,15 +81,21 @@ public String startDependencyContainer(String image, String[] envVars, String po
...
+	public String startContainer(String image, String portMapping, String network, String containerName, Map<String, String> envVars,
+	                             boolean dependencyContainer) throws IOException, InterruptedException {
 		Preconditions.checkNotNull(image, "image");
 		Preconditions.checkNotNull(portMapping, "portMapping");
 
 		String localIp = InetAddress.getLocalHost().getHostAddress();
-		List<String> cmd = new ArrayList<>(Arrays.asList(dockerExePath, "run", "-d", "-p", portMapping, "--add-host=fakeingestion:"+localIp));
+		List<String> cmd = new ArrayList<>();
+		cmd.addAll(Arrays.asList(dockerExePath, "run", "-d", "-p", portMapping, "--add-host=fakeingestion:"+localIp));
+		if (!dependencyContainer) {
```
Comment (littleaj) [inline test/smoke/framework/testCore/src/main/java/com/microsoft/applicationinsights/smoketest/docker/AiDockerClient.java]: nice!

File: agent/src/main/java/com/microsoft/applicationinsights/agent/internal/AIAgentXmlLoader.java [hunk 6]
```diff
@@ -50,6 +50,7 @@ static AgentConfiguration load(File agentJarParentFile) {
         for (InstrumentationDescriptor instrumentationDescriptor : InstrumentationDescriptors.read()) {
             switch (instrumentationDescriptor.name()) {
                 case "apache-http-client":
+                case "http-url-connection":
```
Comment (littleaj) [inline agent/src/main/java/com/microsoft/applicationinsights/agent/internal/AIAgentXmlLoader.java]: are there smoke tests for this?

File: agent/src/main/java/com/microsoft/applicationinsights/agent/internal/model/ThreadContextImpl.java [hunk 7]
```diff
@@ -8,15 +8,14 @@
...
 import org.checkerframework.checker.nullness.qual.Nullable;
 import org.glowroot.xyzzy.engine.bytecode.api.ThreadContextPlus;
-import org.glowroot.xyzzy.engine.bytecode.api.ThreadContextThreadLocal;
 import org.glowroot.xyzzy.engine.impl.NopTransactionService;
 import org.glowroot.xyzzy.instrumentation.api.*;
 
 import java.util.concurrent.TimeUnit;
 
-public class ThreadContextImpl implements ThreadContextPlus {
+import static com.google.common.base.Preconditions.checkNotNull;
```
[the remark this answers is not in the document]
Comment (littleaj) [reply agent/src/main/java/com/microsoft/applicationinsights/agent/internal/model/ThreadContextImpl.java]: discussed and we're going to use the latest version which supports java 7 and has the CVE fix

File: agent/src/main/java/com/microsoft/applicationinsights/agent/internal/MainEntryPoint.java (hunk 4 above)
Comment (trask) [reply agent/src/main/java/com/microsoft/applicationinsights/agent/internal/MainEntryPoint.java]: This code is not called from anywhere yet (it's a possible-feature-in-progress)

**Pull request merged**